android/java/org/chromium/chrome/browser/vpn/billing/InAppPurchaseWrapper.java

```diff
@@ -286,4 +330,52 @@ private PurchasesUpdatedListener getPurchasesUpdatedListener(Context context) {
             }
         };
     }
+
+    private int maxTries;
+    private int tries;
+    private boolean isConnectionEstablished;
+    private void retryBillingServiceConnection(
+            MutableLiveData<Boolean> billingClientConnectionState) {
+        maxTries = 3;
+        tries = 1;
+        isConnectionEstablished = false;
+        do {
+            try {
+                // End existing connection if any before we start another connection
+                endConnection();
+
+                Context context = ContextUtils.getApplicationContext();
+
+                mBillingClient = BillingClient.newBuilder(context)
```

Comment:
it's good to combine that object creation with `startBillingServiceConnection `. Good for follow up as we are in rush right now. I just don't want if we add something new in listener, we need to duplicate it.

Reply:
Yeah that's the reason, I kept it separate. I have verified the that the retry would work only 3 times and then it would stop